README
Crypt-U2F version 0.01
======================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2014 by Rene Schickbauer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Crypt::U2F.

0.10  Wed Dec 17 17:00:00 2014
	- Initial alpha release


u2f.c
/* =====================================================
 * Adapted as a Perl library by Rene 'cavac' Schickbauer
 *
 * This roughly based on u2f-server.c from Yubico's
 * C library, see https://developers.yubico.com/libu2f-server/
 *
 * In order for this to work, you need to install that
 * library.
 *
 * This adaption is (C) 2014 Rene 'cavac' Schickbauer, but as it
 * is based on Yubico's code, the licence below applies!
 *
 * We, the community, would hereby thank Yubico for open
 * sourcing their code!
 * ======================================================
 */
/*
* Copyright (c) 2014 Yubico AB
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* * Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* * Redistributions in binary form must reproduce the above
* copyright notice, this list of conditions and the following
* disclaimer in the documentation and/or other materials provided
* with the distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <u2f-server/u2f-server.h>
#include "u2f.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
//#include <stdbool.h>
#include <string.h>

char *p;
char errorstring[10000];
char regdata[20000];

char* u2fclib_getError(void) {
    return errorstring;
}

int u2fclib_init(int debug) {
  u2fs_rc rc;

  errorstring[0] = 0; // init with empty string

  rc = u2fs_global_init(debug ? U2FS_DEBUG : 0);
  if (rc != U2FS_OK) {
    sprintf(errorstring, "error: u2fs_global_init (%d): %s\n", rc,
            u2fs_strerror(rc));
    return 0;
  }
  return 1;
}

void* u2fclib_get_context(void) {
  u2fs_rc rc;
  u2fs_ctx_t *ctx;

  rc = u2fs_init(&ctx);
  if (rc != U2FS_OK) {
    sprintf(errorstring, "error: u2fs_init (%d): %s\n", rc, u2fs_strerror(rc));
    return 0;
  }

  return (void *)ctx;
}

int u2fclib_setKeyHandle(void* ctx, char* buf) {
  u2fs_rc rc;

    rc = u2fs_set_keyHandle(ctx, buf);
    if (rc != U2FS_OK) {
      sprintf(errorstring, "error: u2fs_set_keyHandle (%d): %s\n", rc,
              u2fs_strerror(rc));
      return 0;
    }

    return 1;
}


int u2fclib_setPublicKey(void* ctx, char* buf) {
  u2fs_rc rc;

    rc = u2fs_set_publicKey(ctx, (unsigned char *) buf);
    if (rc != U2FS_OK) {
      sprintf(errorstring, "error: u2fs_set_publicKey (%d): %s\n", rc,
              u2fs_strerror(rc));
      return 0;
    }

    return 1;
}

int u2fclib_setOrigin(void* ctx, char* origin) {
  u2fs_rc rc;

  rc = u2fs_set_origin(ctx, origin);
  if (rc != U2FS_OK) {
    printf("error: u2fs_set_origin (%d): %s\n", rc, u2fs_strerror(rc));
    return 0;
  }
  return 1;
}

int u2fclib_setAppID(void* ctx, char* appid) {
  u2fs_rc rc;

  rc = u2fs_set_appid(ctx, appid);
  if (rc != U2FS_OK) {
    sprintf(errorstring, "error: u2fs_set_appid (%d): %s\n", rc,
            u2fs_strerror(rc));
    return 0;
  }
  return 1;
}

int u2fclib_setChallenge(void* ctx, char* challenge) {
  u2fs_rc rc;

    rc = u2fs_set_challenge(ctx, challenge);
    if (rc != U2FS_OK) {
      sprintf(errorstring, "error: u2fs_set_challenge (%d): %s\n", rc,
              u2fs_strerror(rc));
      return 0;
    }
    return 1;
}

char* u2fclib_calcRegistrationChallenge(void* ctx) {
  u2fs_rc rc;

    rc = u2fs_registration_challenge(ctx, &p);
    if (rc != U2FS_OK) {
        sprintf(errorstring, "error (%d): %s\n", rc, u2fs_strerror(rc));
        p[0] = 0;
    }
    return p;
}

char* u2fclib_calcAuthenticationChallenge(void* ctx) {
  u2fs_rc rc;

    rc = u2fs_authentication_challenge(ctx, &p);
    if (rc != U2FS_OK) {
        sprintf(errorstring, "error (%d): %s\n", rc, u2fs_strerror(rc));
        p[0] = 0;
    }
    return p;
}

char* u2fclib_verifyRegistration(void* ctx, char* buf) {
  u2fs_rc rc;
    u2fs_reg_res_t *reg_result;

    rc = u2fs_registration_verify(ctx, buf, &reg_result);
    if (rc != U2FS_OK) {
        sprintf(errorstring, "error (%d): %s\n", rc, u2fs_strerror(rc));
      return 0;
    }
    const char *k = u2fs_get_registration_publicKey(reg_result);
    memcpy(regdata, k, U2FS_PUBLIC_KEY_LEN);
    regdata[U2FS_PUBLIC_KEY_LEN] = 0;

    strcat(regdata, "##DELIMITER##");
    strcat(regdata, u2fs_get_registration_keyHandle(reg_result));

    //sprintf(regdata.keyHandle, "%s", u2fs_get_registration_keyHandle(reg_result));

    return regdata;
}

int u2fclib_verifyAuthentication(void* ctx, char* buf) {
  u2fs_rc rc;
    u2fs_auth_res_t *auth_result;

    rc = u2fs_authentication_verify(ctx, buf, &auth_result);
    if (rc == U2FS_OK) {
      u2fs_rc verified;
      uint32_t counter;
      uint8_t user_presence;
      rc = u2fs_get_authentication_result(auth_result, &verified, &counter,
                                          &user_presence);
      if (verified == U2FS_OK) {
        return 1;
      } else {
        sprintf(errorstring, "Authentication failed: %s\n", u2fs_strerror(rc));
        return 0;
      }
    } else if (rc != U2FS_OK) {
      sprintf(errorstring, "error: u2fs_authentication_verify (%d): %s\n", rc,
              u2fs_strerror(rc));
      return 0;
    }
}

int u2fclib_free_context(void* ctx) {
  u2fs_done(ctx);
  return 1;
}

int u2fclib_deInit(void) {
  u2fs_global_done();
  return 1;
}


u2f.h
/* =====================================================
 * Adapted as a Perl library by Rene 'cavac' Schickbauer
 *
 * This roughly based on u2f-server.c from Yubico's
 * C library, see https://developers.yubico.com/libu2f-server/
 *
 * In order for this to work, you need to install that
 * library.
 *
 * This adaption is (C) 2014 Rene 'cavac' Schickbauer, but as it
 * is based on Yubico's code, the licence below applies!
 *
 * We, the community, would hereby thank Yubico for open
 * sourcing their code!
 * ======================================================
 */
/*
* Copyright (c) 2014 Yubico AB
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* * Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* * Redistributions in binary form must reproduce the above
* copyright notice, this list of conditions and the following
* disclaimer in the documentation and/or other materials provided
* with the distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

char* u2fclib_getError(void);
int u2fclib_init(int debug);
void* u2fclib_get_context(void);
int u2fclib_setKeyHandle(void* ctx, char* buf);
int u2fclib_setPublicKey(void* ctx, char* buf);
int u2fclib_setOrigin(void* ctx, char* origin);
int u2fclib_setAppID(void* ctx, char* appid);
int u2fclib_setChallenge(void* ctx, char* challenge);
char* u2fclib_calcRegistrationChallenge(void* ctx);
char* u2fclib_calcAuthenticationChallenge(void* ctx);
char* u2fclib_verifyRegistration(void* ctx, char* buf);
int u2fclib_verifyAuthentication(void* ctx, char* buf);
int u2fclib_free_context(void* ctx);
int u2fclib_deInit(void);


Makefile.PL
use 5.018001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Crypt::U2F',
    VERSION_FROM      => 'lib/Crypt/U2F.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Crypt/U2F.pm', # retrieve abstract from module
       AUTHOR         => 'Rene Schickbauer <cavac@(none)>') : ()),
    LIBS              => ['-lu2f-server'], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
    OBJECT            => '$(O_FILES)', # link all the C files too
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Crypt::U2F',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


lib/Crypt/U2F.pm
package Crypt::U2F;

use 5.018001;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);
our $VERSION = '0.10';

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Crypt::U2F ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	u2fclib_calcAuthenticationChallenge
	u2fclib_calcRegistrationChallenge
	u2fclib_deInit
	u2fclib_free_context
	u2fclib_getError
	u2fclib_get_context
	u2fclib_init
	u2fclib_setAppID
	u2fclib_setChallenge
	u2fclib_setKeyHandle
	u2fclib_setOrigin
	u2fclib_setPublicKey
	u2fclib_verifyAuthentication
	u2fclib_verifyRegistration
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&Crypt::U2F::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#XXX	if ($] >= 5.00561) {
#XXX	    *$AUTOLOAD = sub () { $val };
#XXX	}
#XXX	else {
	    *$AUTOLOAD = sub { $val };
#XXX	}
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('Crypt::U2F', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Crypt::U2F - Low level wrapper around the U2F two factor athentication C library (server side)

=head1 SYNOPSIS

  use Crypt::U2F;

=head1 DESCRIPTION

This is a very low level wrapper around the original C library. You probably B<shouldn't> use it,
but use L<Crypt::U2F::Simple> instead!

This API is subject to change, depending on the underlying library, the weather and the whims of
the developer.

If you decide to use it anyway, it would probably be a good idea to specify the exact version number
of L<Crypt::U2F> to use.

=head1 INSTALLATION

This module requires the Yubico u2f-server shared libary installed, please see the official
project page at L<https://developers.yubico.com/libu2f-server/> on how to do that.

=head1 NO MULTITHREADING / MULTI INSTANCES

The way this is currently implemented, i doubt very much that multithreading or even use in
more than one instance in your program will work. Multi-Forking should be OK, though, if you
only call u2fclib_init() after forking, but this isn't tested as of yet.

The problem is how u2fclib_init() and u2fclib_deInit() are implemented. While the underlying
library has a context-handle (not used here), u2fclib_deInit() will tear down everything as far
as i can tell.

L<Crypt::U2F::Simple> works around this problem by reference counting and always requiring all
values to call the relevant u2fclib_set* functions.

=head2 EXPORT

None by default.

=head2 Exportable functions


  char* u2fclib_calcRegistrationChallenge(void* ctx)
  int u2fclib_deInit(void)
  int u2fclib_free_context(void* ctx)
  char* u2fclib_getError(void)
  void* u2fclib_get_context(void)
  int u2fclib_init(bool debug)
  int u2fclib_setAppID(void* ctx, char* appid)
  int u2fclib_setChallenge(void* ctx, char* challenge)
  int u2fclib_setKeyHandle(void* ctx, char* buf)
  int u2fclib_setOrigin(void* ctx, char* origin)
  int u2fclib_setPublicKey(void* ctx, char* buf)
  int u2fclib_verifyAuthentication(void* ctx, char* buf)
  registrationData_t* u2fclib_verifyRegistration(void* ctx, char* buf)



=head1 SEE ALSO

See L<Crypt::U2F::Simple> for the module you should actually be using.

=head1 BUGS

Yes, there should be some in there. First of all, this is crypto stuff, so
it's broken by default (it only depends on the time it takes to happen).

Also, at the moment, this module has seen only very limited testing.

=head1 AUTHOR

Rene Schickbauer, E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head1 COPYRIGHT AND LICENSE

Adapted as a Perl library by Rene 'cavac' Schickbauer

This roughly based on u2f-server.c from Yubico's
C library, see L<https://developers.yubico.com/libu2f-server/>

In order for this to work, you need to install that
library.

This adaption is (C) 2014 Rene 'cavac' Schickbauer, but as it
is based on Yubico's code, the licence below applies!

I<We, the community, would hereby thank Yubico for open sourcing their code!>

    /*
    * Copyright (c) 2014 Yubico AB
    * All rights reserved.
    *
    * Redistribution and use in source and binary forms, with or without
    * modification, are permitted provided that the following conditions are
    * met:
    *
    * * Redistributions of source code must retain the above copyright
    * notice, this list of conditions and the following disclaimer.
    *
    * * Redistributions in binary form must reproduce the above
    * copyright notice, this list of conditions and the following
    * disclaimer in the documentation and/or other materials provided
    * with the distribution.
    *
    * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
    * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
    * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
    * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
    * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
    * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
    */

=cut


examples/u2fsimple_register.pl
#!/usr/bin/env perl

use strict;
use warnings;

BEGIN {
    unshift @INC, "../lib";
}

my $u2fhost = '/usr/local/bin/u2f-host';
my $appId = 'Example';
my $origin = 'http://127.0.0.1';

use Crypt::U2F::Simple;
use MIME::Base64;

my $auth = Crypt::U2F::Simple->new(appId=>$appId, origin=>$origin);
if(!defined($auth)) {
    die(Crypt::U2F::Simple::lastError());
}

my $challenge = $auth->registrationChallenge();
if(!defined($challenge) || !length($challenge)) {
    die($auth->lastError());
}
open(my $cofh, '>', 'regChallenge.dat') or die($!);
print $cofh $challenge;
close $cofh;

my $regcmd = $u2fhost . ' -aregister -o "' . $origin . '" < regChallenge.dat > regReply.dat';
print "Running $regcmd...\nPlease press the blinking button!\n";
`$regcmd`;

open(my $cifh, '<', 'regReply.dat') or die($!);
my $reply = <$cifh>;
close $cifh;

print "Got $reply\n";

my ($keyHandle, $publicKey) = $auth->registrationVerify($reply);

if(!defined($keyHandle)) {
    print "failed to get keyHandle!\n";
}

if(!defined($publicKey)) {
    print "failed to get publicKey!\n";
}

if(!defined($keyHandle) || !defined($publicKey)) {
    die($auth->lastError());
}

open(my $kofh, '>', 'keyHandle.dat') or die($!);
print $kofh $keyHandle;
close $kofh;

open(my $pofh, '>', 'publicKey.dat') or die($!);
print $pofh encode_base64($publicKey, '');
close $pofh;


lib/Crypt/U2F/Simple.pm
package Crypt::U2F::Simple;

use 5.018001;
use strict;
use warnings;
use Carp;

require Exporter;

our @ISA = qw(Exporter);
our $VERSION = '0.10';

use Crypt::U2F;

my $refCount = 0;
my $errstr = '';

our %EXPORT_TAGS = ( 'all' => [ qw() ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();

my $maxStringLength = 10_000;

sub new {
    my ($proto, %config) = @_;
    my $class = ref($proto) || $proto;

    foreach my $key (qw[appId origin]) {
        if(!defined($config{$key})) {
            croak("$key not defined!");
        }
        if(!length($config{$key})) {
            croak("$key must not be empty!");
        }
    }

    my $self = bless \%config, $class;

    if(!$refCount) {
        my $rc = Crypt::U2F::u2fclib_init(0);
        if(!$rc) {
            $errstr = Crypt::U2F::u2fclib_getError();
            return;
        }
        $refCount++;
    }

    print "Test fail: ", Crypt::U2F::u2fclib_getError(), "\n";

    my $ctx = Crypt::U2F::u2fclib_get_context();
    if(!defined($ctx) || !$ctx) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return;
    }
    $self->{ctx} = $ctx;

    {
        my $rc = Crypt::U2F::u2fclib_setAppID($self->{ctx}, $self->{appId});
        if(!$rc) {
            $errstr = Crypt::U2F::u2fclib_getError();
            return;
        }
    }

    {
        my $rc = Crypt::U2F::u2fclib_setOrigin($self->{ctx}, $self->{origin});
        if(!$rc) {
            $errstr = Crypt::U2F::u2fclib_getError();
            return;
        }
    }

    if(defined($self->{keyHandle})) {
        my $rc = $self->setKeyHandle;
        if(!$rc) {
            $errstr = Crypt::U2F::u2fclib_getError();
            return;
        }
    }

    if(defined($self->{publicKey})) {
        my $rc = $self->setPublicKey;
        if(!$rc) {
            $errstr = Crypt::U2F::u2fclib_getError();
            return;
        }
    }

    return $self;
}

sub lastError {
    return $errstr;
}

sub registrationChallenge {
    my ($self) = @_;

    my $rc = Crypt::U2F::u2fclib_calcRegistrationChallenge($self->{ctx});
    if(!defined($rc) || !length($rc)) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return;
    }

    return $rc;
}

sub registrationVerify {
    my ($self, $registration) = @_;

    if(length($registration) >= $maxStringLength) {
        $errstr = "Registration string too long!";
        return;
    }

    my $keydata = Crypt::U2F::u2fclib_verifyRegistration($self->{ctx}, $registration);
    if(!defined($keydata)) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return;
    }

    ($self->{publicKey}, $self->{keyHandle}) = split/\#\#DELIMITER\#\#/, $keydata;

    return ($self->{keyHandle}, $self->{publicKey});
}

sub setKeyHandle {
    my ($self, $keyHandle) = @_;

    if(!defined($keyHandle) && !defined($self->{keyHandle})) {
        $errstr = "No keyHandle given!";
        return 0;
    }
    if(defined($keyHandle)) {
        $self->{keyHandle} = $keyHandle;
    }

    if(length($self->{keyHandle}) >= $maxStringLength) {
        $errstr = "keyHandle string too long!";
        return 0;
    }

    my $rc = Crypt::U2F::u2fclib_setKeyHandle($self->{ctx}, $self->{keyHandle});
    if(!defined($rc) || !$rc) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return;
    }

    return 1;
}

sub setPublicKey {
    my ($self, $publicKey) = @_;

    if(!defined($publicKey) && !defined($self->{publicKey})) {
        $errstr = "No publicKey given!";
        return 0;
    }
    if(defined($publicKey)) {
        $self->{publicKey} = $publicKey;
    }

    if(length($self->{publicKey}) >= $maxStringLength) {
        $errstr = "publicKey string too long!";
        return 0;
    }

    my $rc = Crypt::U2F::u2fclib_setPublicKey($self->{ctx}, $self->{publicKey});
    if(!defined($rc) || !$rc) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return;
    }

    return 1;
}

sub authenticationChallenge {
    my ($self) = @_;

    print "BLI\n";
    if(!$self->setKeyHandle || !$self->setPublicKey) {
        return;
    }

    my $rc = Crypt::U2F::u2fclib_calcAuthenticationChallenge($self->{ctx});
    print "BLUB\n";
    if(!defined($rc) || !length($rc)) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return;
    }

    return $rc;
}

sub authenticationVerify {
    my ($self, $authentication) = @_;

    if(!$self->setKeyHandle || !$self->setPublicKey) {
        return 0;
    }

    my $rc = Crypt::U2F::u2fclib_verifyAuthentication($self->{ctx}, $authentication);
    if(!defined($rc) || !$rc) {
        $errstr = Crypt::U2F::u2fclib_getError();
        return 0;
    }

    return $rc;
}

sub DESTROY {
    my ($self) = @_;

    Crypt::U2F::u2fclib_free_context($self->{ctx});

    $refCount--;
    if($refCount < 0) {
        croak("refCount error! Aborting everything!");
    }

    if(!$refCount) {
        Crypt::U2F::u2fclib_deInit();
    }

    return;
}


1;
__END__

=head1 NAME

Crypt::U2F::Simple - Register and Authenticate U2F compatible security devices

=head1 SYNOPSIS

  use Crypt::U2F:Simple;

  my $crypter = Crypt::U2F::Simple->new(appId => 'Perl', origin='http://search.cpan.org');

  # Generate a registration request
  my $registerRequest = $crypter->registrationChallenge();

  # Give $registerRequest to client, recieve $registrationData from client
  my ($keyHandle, $userKey) = $crypter->registrationVerify($registrationData)

  # Generate an authentication request (using the previously generated key handle and user key)
  my $authrequest = $crypter->authenticationChallenge();

  # Send $authrequest to client, recieve $authSignature
  my $authok = $crypter->authenticationVerify($authSignature);

=head1 DESCRIPTION

This module implements the server side of U2F authentication through Yubico's C library.

Both registration and authentication are two step processes that each must be run in the same instance
of this perl module. To clarify You can run registration from another instance than authentication, or even
in another program on another server. But, as far as it is currently implemented, you must run both registration
steps in the same instance of this module, the same goes for authentication. Needs more testing, really.

A succesfull registration of a key yields to two scalars, a key handle and a public key. It is B<your>
responsibility to keep them safe somewhere and reload them into this module whenever you want to do
authentication.

=head1 INSTALLATION

This module requires the Yubico u2f-server shared libary installed, please see the official
project page at L<https://developers.yubico.com/libu2f-server/> on how to do that.

=head1 NO MULTITHREADING

The way this is currently implemented, i doubt very much that multithreadingm will work.
Multi-Forking should be OK as long as you only call new() B<after> forking, though. Also
using more than one instance of this module in your program. This isn't really tested 
at the moment, though...

=head1 ALPHA WARNING

As already stated above, at this time L<Crypt::U2F> and L<Crypt::U2F::Simple> have seen only
very limited testing and the modules are still subject to change.

That isn't to say that you shouldn't use this at all. Rather, if you are interested, you should test this
a lot and report any bugs you find!

=head1 FUNCTION DESCRIPTION

=head2 lastError()

Probably the most important function of all, therefore mention first. If something goes wrong
(but not HorriblyWrong[tm]), you'll get the last error description of whatever happened.

If something fails during new(), call it with the full name:

    my $oooops = Crypt::U2F::Simple::lastError();

If you already got an instance, you can use that as well:

    my $oooops = $auth->lastError();

Errors are global over all instances of this module.

If things go HorriblyWrong[tm], your program might crash. Or get remote-buffer-overflow-exploited
or something. In these case, lastError() might not work reliably. You know, just the usual crypto
stuff...

=head2 new()

This comes in two forms, depending if you do authentication in the same instance as the registration steps.

The simple form (only registration or registration+authentication) only requires the arguments appId and origin:

    my $auth = Crypt::U2F::Simple->new(appId => 'Perl', 
                                       origin='http://search.cpan.org');

If you only do authentication, you have to supply the keyHandle and publicKey data as well:


    my $auth = Crypt::U2F::Simple->new(appId => 'Perl', 
                                       origin='http://search.cpan.org',
                                       keyHandle=$keyHandleData
                                       publicKey=$publicKeyData);

For security, i would recommend creating a new instance for each and every authentication request.

If something goes wrong during initialization, $auth will be I<undef>.

=head2 registrationChallenge()

    my $challenge = $auth->registrationChallenge();

Gives you a unique registration challenge on every call. This is a JSON string and should be send to the client
(called a "host" for whatever reason) as is.

If something goes wrong, $challenge will be I<undef>.

=head2 registrationVerify()

    my ($keyHandle, $publicKey) = $auth->registrationVerify($reply);

If the client (the "host") accepts the challenge, it will send you another JSON blob ($reply). 

If everything goes well and registration succeeds, you will get the key handle and public key of, well
client key. If it fails, you will get I<undef>.

$keyHandle and $publicKey will get set internally for direct following authentication in the same instance,
you need to store it in some persistent way yourself for future authentication.

As an added bonus, $publicKey will be a binary blob, so you may have to convert it to something like Base64 for
easier handling. See L<MIME::Base64> on how to do that. Make sure you un-encode before loading it into this
module!

=head2 authenticationChallenge()

This function generates an authentification challenge. To do that, it needs keyHandle and publicKey, since this
is key dependend.

    my $challenge = $auth->authenticationChallenge();

Otherwise, this works the same as the registration challenge. You get a JSON blob, send that to the client and
get an answer.

=head2 authenticationVerify()

After you get the authentication answer, you need to verify it:

    my $isValid = $auth->authenticationVerify($reply);

$isValid is true if authentication succeedsr. If something went wrong (library error, fake user), $isValid is false, in which
case you can look into lastError() to see what went wrong.

=head1 SEE ALSO

See L<Crypt::U2F> for the low level library if you want better headaches.

There are two examples in the tarball for registration and authentication.

=head1 BUGS

Yes, there should be some in there. First of all, this is crypto stuff, so
it's broken by default (it only depends on the time it takes to happen).

Also, at the moment, this module has seen only very limited testing.

=head1 AUTHOR

Rene Schickbauer, E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head1 COPYRIGHT AND LICENSE

Adapted as a Perl library by Rene 'cavac' Schickbauer

This roughly based on u2f-server.c from Yubico's
C library, see L<https://developers.yubico.com/libu2f-server/>

In order for this to work, you need to install that
library.

This adaption is (C) 2014 Rene 'cavac' Schickbauer, but as it
is based on Yubico's code, the licence below applies!

I<We, the community, would hereby thank Yubico for open sourcing their code!>

    /*
    * Copyright (c) 2014 Yubico AB
    * All rights reserved.
    *
    * Redistribution and use in source and binary forms, with or without
    * modification, are permitted provided that the following conditions are
    * met:
    *
    * * Redistributions of source code must retain the above copyright
    * notice, this list of conditions and the following disclaimer.
    *
    * * Redistributions in binary form must reproduce the above
    * copyright notice, this list of conditions and the following
    * disclaimer in the documentation and/or other materials provided
    * with the distribution.
    *
    * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
    * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
    * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
    * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
    * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
    * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
    */

=cut
